auto_spark_session.py

```diff
@@ -0,0 +1,168 @@
+# NOTE: There can be multiple client secret keys for a service account.
+# NOTE: service principal requires "Storage Blob Data Contributor" role on storage account.
+# NOTE: application_id is found here: MS Entra admin center > Applications > App Registrations > Overview: Application (client) ID
+# NOTE: directory_id is found here: MS Entra admin center > Applications > App Registrations > Overview: Directory (tenent) ID
+
+
+
+# OK, how this works is this:
+# When we create a SparkConf, that is the Python side representation of the config.
+# When we call SparkSession.builder() with the SparkConf, it creates the JVM and
+# all the relevant JavaObjects containing the Scala side representation of the config.
+# Attached to the SparkSession is a RuntimeConfig object which acts as the API for
+# for Python to get/set config options with the JavaObject during runtime. So, what we
+# do is to intercept any calls to this RuntimeConfig, mask any containing secrets, and
+# passing any other calls to the API without modification.
+
+# XXX Wondering if we can do something neat like encrypt the file with a cert?
+
+class WrapRuntimeConfig(RuntimeConfig):
+    """RuntimeConfig interface wrapper
+
+    When a user makes a request to the RuntimeConfig for a key which is likely to contain
+    a secret, we mask the value returned. All other requests are passed unaltered.
+    """
+    def __init__(self, *args: Any, **kwargs: Any):
+        super().__init__(*args, **kwargs)
+        self._mask_keys = ["secret", "password", "passwd"]
+        self._mask = "*******"
+
+    # XXX How to handle version differences?
+    def get(self, key: str, defaultValue: Union[Optional[str], _NoValueType] = _NoValue) -> Optional[str]:
+        if any(m in key for m in self._mask_keys):
+            return f"{self._mask}"
+        return super().get(key, defaultValue)
+
+    # XXX I noticed getAll() was available on newer versions
+
+
+
+class ServicePrincipal:
+    # XXX Is this the right place for this?
+    _basedir_path = "/tmp/stuff/service_principals"
+    # XXX check if we can do the following:
+    # 1. Have a /path/to/some/HASH/sp_dir
+    # 2. Change the ownership to root from /path/to and down to the leaves
+    # 3. Then chmod 640 as well.
+    # 4. Then see if we can read the absolute path, if we know it, but cannot actually
+    #    browse to it in a directory.
+
+    def __init__(
+        self,
+        name: str,
+        client_secret_id: str,
+        client_secret_value: str,
+        application_id: str,
+        directory_id: str,
+        expires: str
+    ) -> None:
+        self.name = name
+        self.client_secret_id = client_secret_id
+        self.client_secret_value = client_secret_value
+        self.application_id = application_id
+        self.directory_id = directory_id
+        self.expires = expires
+
+    @classmethod
+    def from_files(cls, name: str) -> "ServicePrincipal":
```

Comment:
Cause for the confusion above, would be nice to comment/docstring or rename what `name` is here, since there is a `self.name`

Reply:
Changed to `key`